### PR TITLE
Skyline syntax (preferring clause)

### PR DESCRIPTION
**Pull request by ssteinhauser**
This PR implements the skyline syntax (preferring clause) of Exasol which is available for `SELECT`, `UPDATE` and `DELETE` statements. For more details see https://docs.exasol.com/db/latest/advanced_analytics/skyline.htm

### CODE REVIEW

**Comment by manticore-projects** [comment]
Thank you much for your contribution and effort!
Are you planing more PRs like this? I ask to time a possible 5.1 release with you which would carry your work.

File: src/main/java/net/sf/jsqlparser/expression/ExpressionVisitor.java [hunk 1]
```diff
@@ -24,34 +24,7 @@
...
-import net.sf.jsqlparser.expression.operators.relational.Matches;
-import net.sf.jsqlparser.expression.operators.relational.MemberOfExpression;
-import net.sf.jsqlparser.expression.operators.relational.MinorThan;
-import net.sf.jsqlparser.expression.operators.relational.MinorThanEquals;
-import net.sf.jsqlparser.expression.operators.relational.NotEqualsTo;
-import net.sf.jsqlparser.expression.operators.relational.RegExpMatchOperator;
-import net.sf.jsqlparser.expression.operators.relational.SimilarToExpression;
-import net.sf.jsqlparser.expression.operators.relational.TSQLLeftJoin;
-import net.sf.jsqlparser.expression.operators.relational.TSQLRightJoin;
+import net.sf.jsqlparser.expression.operators.relational.*;
```
Comment (manticore-projects) [inline src/main/java/net/sf/jsqlparser/expression/ExpressionVisitor.java:27]: Please, no wild card imports!

File: src/main/java/net/sf/jsqlparser/statement/select/PlainSelect.java [hunk 2]
```diff
@@ -9,11 +9,7 @@
  */
 package net.sf.jsqlparser.statement.select;
 
-import net.sf.jsqlparser.expression.Alias;
-import net.sf.jsqlparser.expression.Expression;
-import net.sf.jsqlparser.expression.OracleHierarchicalExpression;
-import net.sf.jsqlparser.expression.OracleHint;
-import net.sf.jsqlparser.expression.WindowDefinition;
+import net.sf.jsqlparser.expression.*;
```
Comment (manticore-projects) [inline src/main/java/net/sf/jsqlparser/statement/select/PlainSelect.java:12]: Please no wild card imports

File: src/main/java/net/sf/jsqlparser/util/TablesNamesFinder.java [hunk 3]
```diff
@@ -10,59 +10,7 @@
...
-import net.sf.jsqlparser.expression.TimeValue;
-import net.sf.jsqlparser.expression.TimestampValue;
-import net.sf.jsqlparser.expression.TimezoneExpression;
-import net.sf.jsqlparser.expression.TranscodingFunction;
-import net.sf.jsqlparser.expression.TrimFunction;
-import net.sf.jsqlparser.expression.UserVariable;
-import net.sf.jsqlparser.expression.VariableAssignment;
-import net.sf.jsqlparser.expression.WhenClause;
-import net.sf.jsqlparser.expression.XMLSerializeExpr;
+import net.sf.jsqlparser.expression.*;
```
Comment (manticore-projects) [inline src/main/java/net/sf/jsqlparser/util/TablesNamesFinder.java:13]: Please no wild card imports.

File: src/main/java/net/sf/jsqlparser/util/TablesNamesFinder.java [hunk 4]
```diff
@@ -78,34 +26,7 @@
...
-import net.sf.jsqlparser.expression.operators.relational.Matches;
-import net.sf.jsqlparser.expression.operators.relational.MemberOfExpression;
-import net.sf.jsqlparser.expression.operators.relational.MinorThan;
-import net.sf.jsqlparser.expression.operators.relational.MinorThanEquals;
-import net.sf.jsqlparser.expression.operators.relational.NotEqualsTo;
-import net.sf.jsqlparser.expression.operators.relational.RegExpMatchOperator;
-import net.sf.jsqlparser.expression.operators.relational.SimilarToExpression;
-import net.sf.jsqlparser.expression.operators.relational.TSQLLeftJoin;
-import net.sf.jsqlparser.expression.operators.relational.TSQLRightJoin;
+import net.sf.jsqlparser.expression.operators.relational.*;
```
Comment (manticore-projects) [inline src/main/java/net/sf/jsqlparser/util/TablesNamesFinder.java:29]: Please no wild card imports.

File: src/main/java/net/sf/jsqlparser/util/deparser/ExpressionDeParser.java [hunk 5]
```diff
@@ -9,60 +9,7 @@
...
-import net.sf.jsqlparser.expression.TimestampValue;
-import net.sf.jsqlparser.expression.TimezoneExpression;
-import net.sf.jsqlparser.expression.TranscodingFunction;
-import net.sf.jsqlparser.expression.TrimFunction;
-import net.sf.jsqlparser.expression.UserVariable;
-import net.sf.jsqlparser.expression.VariableAssignment;
-import net.sf.jsqlparser.expression.WhenClause;
-import net.sf.jsqlparser.expression.WindowElement;
-import net.sf.jsqlparser.expression.XMLSerializeExpr;
+import net.sf.jsqlparser.expression.*;
```
Comment (manticore-projects) [inline src/main/java/net/sf/jsqlparser/util/deparser/ExpressionDeParser.java:12]: Please no wild card imports.

File: src/main/java/net/sf/jsqlparser/util/deparser/ExpressionDeParser.java [hunk 6]
```diff
@@ -78,36 +25,7 @@
...
-import net.sf.jsqlparser.expression.operators.relational.MinorThan;
-import net.sf.jsqlparser.expression.operators.relational.MinorThanEquals;
-import net.sf.jsqlparser.expression.operators.relational.NotEqualsTo;
-import net.sf.jsqlparser.expression.operators.relational.OldOracleJoinBinaryExpression;
-import net.sf.jsqlparser.expression.operators.relational.RegExpMatchOperator;
-import net.sf.jsqlparser.expression.operators.relational.SimilarToExpression;
-import net.sf.jsqlparser.expression.operators.relational.SupportsOldOracleJoinSyntax;
-import net.sf.jsqlparser.expression.operators.relational.TSQLLeftJoin;
-import net.sf.jsqlparser.expression.operators.relational.TSQLRightJoin;
+import net.sf.jsqlparser.expression.operators.relational.*;
```
Comment (manticore-projects) [inline src/main/java/net/sf/jsqlparser/util/deparser/ExpressionDeParser.java:28]: Please no wild card imports.

File: src/main/java/net/sf/jsqlparser/util/validation/validator/ExpressionValidator.java [hunk 7]
```diff
@@ -9,61 +9,7 @@
...
-import net.sf.jsqlparser.expression.TranscodingFunction;
-import net.sf.jsqlparser.expression.TrimFunction;
-import net.sf.jsqlparser.expression.UserVariable;
-import net.sf.jsqlparser.expression.VariableAssignment;
-import net.sf.jsqlparser.expression.WhenClause;
-import net.sf.jsqlparser.expression.WindowElement;
-import net.sf.jsqlparser.expression.WindowOffset;
-import net.sf.jsqlparser.expression.WindowRange;
-import net.sf.jsqlparser.expression.XMLSerializeExpr;
+import net.sf.jsqlparser.expression.*;
```
Comment (manticore-projects) [inline src/main/java/net/sf/jsqlparser/util/validation/validator/ExpressionValidator.java:12]: Please no wild card imports.

File: src/main/java/net/sf/jsqlparser/util/validation/validator/ExpressionValidator.java [hunk 8]
```diff
@@ -79,36 +25,7 @@
...
-import net.sf.jsqlparser.expression.operators.relational.MinorThan;
-import net.sf.jsqlparser.expression.operators.relational.MinorThanEquals;
-import net.sf.jsqlparser.expression.operators.relational.NotEqualsTo;
-import net.sf.jsqlparser.expression.operators.relational.OldOracleJoinBinaryExpression;
-import net.sf.jsqlparser.expression.operators.relational.RegExpMatchOperator;
-import net.sf.jsqlparser.expression.operators.relational.SimilarToExpression;
-import net.sf.jsqlparser.expression.operators.relational.SupportsOldOracleJoinSyntax;
-import net.sf.jsqlparser.expression.operators.relational.TSQLLeftJoin;
-import net.sf.jsqlparser.expression.operators.relational.TSQLRightJoin;
+import net.sf.jsqlparser.expression.operators.relational.*;
```
Comment (manticore-projects) [inline src/main/java/net/sf/jsqlparser/util/validation/validator/ExpressionValidator.java:28]: Please no wild card imports.

**Review by manticore-projects — CHANGES_REQUESTED** [review]
Please no wild card imports (as requested by @wumpz).
Rest looks great as usual and I would merge immediately when this will have been  cleaned up. Thanks!

**Comment by manticore-projects** [comment]
Please check for ALL wild card imports (in case I have missed to mark one), thanks!

**Comment by ssteinhauser** [comment]
> Thank you much for your contribution and effort! Are you planing more PRs like this? I ask to time a possible 5.1 release with you which would carry your work.

You're welcome and thank you also for your support!

Currently, I don't have a complete overview of Exasol syntax which is not supported by JSqlParser, but my next implementation will be the support of the [IMPORT ](https://docs.exasol.com/db/latest/sql/import.htm) (with [subimport ](https://docs.exasol.com/db/latest/sql/select.htm#Syntax) as part of select) and [EXPORT ](https://docs.exasol.com/db/latest/sql/export.htm) statements of Exasol.
So a 5.1 release might make sense then.

**Comment by ssteinhauser** [comment]
> Please check for ALL wild card imports (in case I have missed to mark one), thanks!

Will do and sorry - my IDE did that automatically.

**Review by manticore-projects — APPROVED** [review]

**Comment by manticore-projects** [comment]
> > Please check for ALL wild card imports (in case I have missed to mark one), thanks!
> 
> Will do and sorry - my IDE did that automatically.

there should be a configuration for this (even project specific).

**Comment by manticore-projects** [comment]
> > Thank you much for your contribution and effort! Are you planing more PRs like this? I ask to time a possible 5.1 release with you which would carry your work.
> 
> You're welcome and thank you also for your support!
> 
> Currently, I don't have a complete overview of Exasol syntax which is not supported by JSqlParser, but my next implementation will be the support of the [IMPORT ](https://docs.exasol.com/db/latest/sql/import.htm) (with [subimport ](https://docs.exasol.com/db/latest/sql/select.htm#Syntax) as part of select) and [EXPORT ](https://docs.exasol.com/db/latest/sql/export.htm) statements of Exasol. So a 5.1 release might make sense then.

Ok, so I would like to suggest we will wait with 5.1 for this specific stuff and then will aim for a release when this would be ready.
Cheers!

**Pull request merged**